tool/tbot/main.go

```diff
@@ -0,0 +1,607 @@
+			defaultRoles = []string{}
+		}
+
+		// Next, generate impersonated certs
+		expires := time.Unix(int64(ident.Cert.ValidBefore), 0)
+		for _, dest := range cfg.Destinations {
+			destImpl, err := dest.GetDestination()
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			var desiredRoles []string
+			if len(dest.Roles) > 0 {
+				desiredRoles = dest.Roles
+			} else {
+				log.Debugf("Destination specified no roles, defaults will be requested: %v", defaultRoles)
+				desiredRoles = defaultRoles
+			}
+
+			impersonatedIdent, err := generateImpersonatedIdentity(client, ident, expires, desiredRoles)
+			if err != nil {
+				return trace.WrapWithMessage(err, "Failed to generate impersonated certs for %s: %+v", destImpl, err)
+			}
+
+			if err := describeIdentity(impersonatedIdent); err != nil {
+				return trace.WrapWithMessage(err, "could not describe impersonated certs for destination %s", destImpl)
+			}
+
+			if err := identity.SaveIdentity(impersonatedIdent, destImpl); err != nil {
+				return trace.WrapWithMessage(err, "failed to save impersonated identity to destination %s", destImpl)
+			}
+
+			for _, templateConfig := range dest.Configs {
+				template, err := templateConfig.GetConfigTemplate()
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				if err := template.Render(client, impersonatedIdent, dest); err != nil {
+					log.WithError(err).Warnf("Failed to render config template %+v", templateConfig)
+				}
+			}
+		}
+
+		log.Infof("Persisted new certificates to disk. Next renewal in approximately %s", cfg.RenewInterval)
+		<-ticker.C
+	}
+}
+
+func authenticatedUserClientFromIdentity(id *identity.Identity, authServer string) (*auth.Client, error) {
+	tlsConfig, err := id.TLSConfig([]uint16{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	addr, err := utils.ParseAddr(authServer)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return auth.NewClient(api.Config{
+		Addrs: utils.NetAddrsToStrings([]utils.NetAddr{*addr}),
+		Credentials: []api.Credentials{
+			api.LoadTLS(tlsConfig),
+		},
+	})
+}
+
+func generateImpersonatedIdentity(
+	client *auth.Client,
+	currentIdentity *identity.Identity,
+	expires time.Time,
+	roleRequests []string,
+) (*identity.Identity, error) {
+	// TODO: enforce expiration > renewal period (by what margin?)
+
+	// First, ask the auth server to generate a new set of certs with a new
+	// expiration date.
+	certs, err := client.GenerateUserCerts(context.Background(), proto.UserCertsRequest{
+		PublicKey:    currentIdentity.SSHPublicKeyBytes,
+		Username:     currentIdentity.XCert.Subject.CommonName,
+		Expires:      expires,
+		RoleRequests: roleRequests,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The root CA included with the returned user certs will only contain the
+	// Teleport User CA. We'll also need the host CA for future API calls.
+	localCA, err := client.GetClusterCACert()
```

Comment:
for my own understanding, which api calls are made with the impersonated identity?

Reply:
Ideally nothing, unless some end user wants to use the impersonated certs to access Teleport APIs. The "future API calls" referenced in the comment is... copied and pasted from `renewIdentityViaAuth()` so most likely not relevant anymore. These CAs would probably be needed if an end-user wants to access Teleport APIs, though.